pallets/gear-voucher/src/tests.rs

```diff
@@ -1,82 +1,674 @@
+            None,
+            // balance top up
+            None,
+            // extra programs
+            None,
+            // prolong validity
+            Some(validity_prolong),
         ));
 
-        // Redemption fails
+        let voucher = Vouchers::<Test>::get(BOB, voucher_id).unwrap();
+        assert_eq!(voucher.validity, huge_block + validity_prolong);
+    });
+}
+
+#[test]
+fn voucher_update_err_cases() {
+    new_test_ext().execute_with(|| {
+        let program_id = H256::random().cast();
+
+        let voucher_id = utils::issue(ALICE, BOB, program_id);
+        assert_ok!(voucher_id);
+        let voucher_id = voucher_id.unwrap();
+
+        // Inexistent voucher.
+        assert_noop!(
+            Voucher::update(
+                RuntimeOrigin::signed(ALICE),
+                ALICE,
+                voucher_id,
+                // move ownership
+                None,
+                // balance top up
+                None,
+                // extra programs
+                None,
+                // prolong validity
+                None,
+            ),
+            Error::<Test>::InexistentVoucher
+        );
+
+        // Update by non-owner.
+        assert_noop!(
+            Voucher::update(
+                RuntimeOrigin::signed(BOB),
+                BOB,
+                voucher_id,
+                // move ownership
+                None,
+                // balance top up
+                None,
+                // extra programs
+                None,
+                // prolong validity
+                None,
+            ),
+            Error::<Test>::BadOrigin
+        );
+
+        // Balances error.
+        assert_noop!(
+            Voucher::update(
+                RuntimeOrigin::signed(ALICE),
+                BOB,
+                voucher_id,
+                // move ownership
+                None,
+                // balance top up
+                Some(100_000_000_000_000),
+                // extra programs
+                None,
+                // prolong validity
+                None,
+            ),
+            Error::<Test>::BalanceTransfer
+        );
+
+        // Programs limit exceed.
+        let set = (0..=<<Test as Config>::MaxProgramsAmount as Get<u8>>::get())
+            .map(|_| H256::random().cast())
+            .collect();
+
         assert_noop!(
-            Balances::reserve(&Voucher::voucher_id(&BOB, &program_id), 100_000_000),
-            pallet_balances::Error::<Test>::InsufficientBalance
+            Voucher::update(
+                RuntimeOrigin::signed(ALICE),
+                BOB,
+                voucher_id,
+                // move ownership
+                None,
+                // balance top up
+                None,
+                // extra programs
+                Some(Some(set)),
+                // prolong validity
+                None,
+            ),
+            Error::<Test>::MaxProgramsLimitExceeded
         );
     });
 }
+
+mod utils {
```

Comment:
It's a good practice when you read functions from top to bottom (from caller to callee), it really makes increases readability. But here and in `internal.rs` it's a reversed order.